packages/apps/policy-contract-whitelist/src/lib/vincent-policy.ts

```diff
@@ -0,0 +1,145 @@
+import { createVincentPolicy } from '@lit-protocol/vincent-tool-sdk';
+import { ethers } from 'ethers';
+
+import {
+  evalAllowResultSchema,
+  evalDenyResultSchema,
+  precheckAllowResultSchema,
+  precheckDenyResultSchema,
+  toolParamsSchema,
+  userParamsSchema,
+} from './schemas';
+
+export const vincentPolicy = createVincentPolicy({
+  packageName: '@lit-protocol/vincent-policy-contract-whitelist' as const,
+
+  toolParamsSchema,
+  userParamsSchema,
+
+  precheckAllowResultSchema,
+  precheckDenyResultSchema,
+
+  evalAllowResultSchema,
+  evalDenyResultSchema,
+
+  precheck: async ({ toolParams, userParams }, { allow, deny }) => {
+    try {
+      const { serializedTransaction } = toolParams;
+      const { whitelist } = userParams;
+
+      const transaction = ethers.utils.parseTransaction(serializedTransaction);
+      const { chainId, to: contractAddress, data } = transaction;
+      const functionSelector = data.slice(0, 10);
+
+      if (!contractAddress) {
+        return deny({
+          reason: 'to property of serialized transaction not provided',
+        });
+      }
+
+      const chainWhitelist = whitelist[chainId];
+      if (!chainWhitelist) {
+        return deny({
+          reason: 'Chain ID not whitelisted',
+          chainId,
+          contractAddress,
+          functionSelector,
+        });
+      }
+
+      const functionWhitelist = chainWhitelist[contractAddress];
+      if (!functionWhitelist) {
+        return deny({
+          reason: 'Function selector not whitelisted',
+          chainId,
+          contractAddress,
+          functionSelector,
+        });
+      }
+
+      const hasSpecificSelector = functionWhitelist.functionSelectors.includes(functionSelector);
+      const hasWildcard = functionWhitelist.functionSelectors.includes('*');
+
+      if (!hasSpecificSelector && !hasWildcard) {
+        return deny({
+          reason: 'Function selector not whitelisted',
+          chainId,
+          contractAddress,
+          functionSelector,
+        });
+      }
+
+      return allow({
+        chainId,
+        contractAddress,
+        functionSelector,
+        wildcardUsed: hasWildcard && !hasSpecificSelector,
+      });
+    } catch (error) {
+      console.error('Precheck error:', error);
+      return deny({
+        reason: error instanceof Error ? error.message : 'Unknown error',
+      });
+    }
+  },
+  evaluate: async ({ toolParams, userParams }, { allow, deny }) => {
+    try {
+      const { serializedTransaction } = toolParams;
+      const { whitelist } = userParams;
```

Comment:
interesting, they pass in the whitelist?  so we aren't storing it on chain yet, right?  that's next? 

do you think you would add the "pull the whitelist" code into the policy?  or always have the tool get it and pass it in?

Reply:
oh ok works for me